app/controllers/concerns/two_factor_authenticatable.rb

```diff
@@ -33,36 +35,33 @@ def reset_attempt_count_if_user_no_longer_locked_out
     current_user.update(second_factor_attempts_count: 0, second_factor_locked_at: nil)
   end
 
-  def mark_user_session_authenticated
-    user_session[TwoFactorAuthentication::NEED_AUTHENTICATION] = false
-    user_session[:authn_at] = Time.zone.now
-  end
-
   def handle_valid_otp
-    mark_user_session_authenticated
-    bypass_sign_in current_user
-    flash[:notice] = t('devise.two_factor_authentication.success') unless reauthn?
-
-    analytics.track_event('User 2FA successful')
+    if context == 'authentication'
+      handle_valid_otp_for_authentication_context
+    else
+      handle_valid_otp_for_confirmation_context
+    end
 
-    current_user.update(second_factor_attempts_count: 0)
+    redirect_to after_otp_verification_confirmation_path
+  end
 
-    redirect_to after_2fa_path
+  def context
+    context = params[:context]
+    context.present? ? context : 'authentication'
```

Comment:
Do we ever use `.presence`? Could turn this all into:

```
params[:context].presence || 'authentication'
```

which is a little more terse?